distributed/worker.py

```diff
@@ -1966,12 +1940,56 @@ def handle_compute_task(
             "error",
         }:
             recommendations[ts] = "waiting"
-        else:  # pragma: no cover
+
+            ts.run_spec = SerializedTask(function, args, kwargs, task)
+
+            assert isinstance(priority, tuple)
+            priority = priority + (self.generation,)
+            self.generation -= 1
+
+            if actor:
+                self.actors[ts.key] = None
+
+            ts.exception = None
+            ts.traceback = None
+            ts.exception_text = ""
+            ts.traceback_text = ""
+            ts.priority = priority
+            ts.duration = duration
+            if resource_restrictions:
+                ts.resource_restrictions = resource_restrictions
+            ts.annotations = annotations
+
+            if self.validate:
+                assert who_has.keys() == nbytes.keys()
+                assert all(who_has.values())
+
+            for dep_key, dep_workers in who_has.items():
+                dep_ts = self.ensure_task_exists(
+                    key=dep_key,
+                    priority=priority,
+                    stimulus_id=stimulus_id,
+                )
+                # link up to child / parents
+                ts.dependencies.add(dep_ts)
+                dep_ts.dependents.add(ts)
+
+            for dep_key, value in nbytes.items():
+                self.tasks[dep_key].nbytes = value
+
+            self.update_who_has(who_has)
+        else:  # pragma: nocover
             raise RuntimeError(f"Unexpected task state encountered {ts} {stimulus_id}")
 
-        self._handle_instructions(instructions)
-        self.update_who_has(who_has)
         self.transitions(recommendations, stimulus_id=stimulus_id)
+        self._handle_instructions(instructions)
+
+        if self.validate:
+            # All previously unknown tasks that were created above by
+            # ensure_tasks_exists() have been transitioned to fetch or flight
+            assert all(
+                ts2.state != "released" for ts2 in (ts, *ts.dependencies)
+            ), self.story(ts, *ts.dependencies)
```

Comment:
At the moment of writing, this assertion fails in test_stress_scatter_death 0.4% of the times on a fast desktop - meaning there's something else going on.
Resolution TBD.

Reply:
FYI this assert is not 100% correct. There is a case for valid tasks left in released in the case of cancelled/resumed tasks. I'll open a follow up PR with a case reproducing this condition

Reply:
```mermaid
flowchart TD
  A1[A1 - forgotten / not known] --> B1[B1 - flight]
  A2[A1 - forgotten / not known] --> B1[B1 - flight]
  B1 --> C1[C1 - waiting]
```
free-keys / cancel B1
```mermaid
flowchart TD
  A1[A1 - forgotten / not known] --> B1[B1 - cancelled]
  A2[A1 - forgotten / not known] --> B1[B1 - cancelled]
  B1 --> C1[C1 - forgotten]
```

compute-task B1
```mermaid
flowchart TD
  A1[A1 - released] --> B1[B1 - resumed]
  A2[A1 - released] --> B1[B1 - resumed]
  B1 --> C1[C1 - forgotten]
```

gather-dep finishes w/ Error

compute-task B1
```mermaid
flowchart TD
  A1[A1 - fetch] --> B1[B1 - waiting]
  A2[A1 - fetch] --> B1[B1 - waiting]
  B1 --> C1[C1 - forgotten]
```

Reply:
I don't think I understand from the above diagram how this can end up with a released state by the end of compute-task?